### PR TITLE
Jokers

### PR DESCRIPTION
**Pull request by darrencauthon**
Some of my favorite card games have a joker.  I started implementing a "joker" card.  I'm not happy with this so far, as I **hate** `Card.new(nil, nil, 'Red')`

But it's still worth talking about, at least to see if this is worth moving forward.

### CODE REVIEW

**Comment by jdan** [comment]
I think it's a great idea - can Joker be thought of as a rank?

**Comment by darrencauthon** [comment]
I think a joker could be a rank... but I'd hate to have it assumed.

I slept on it, and here are my thoughts... imagine if Cards were initialized like this:

```
card = Card.new(rank: 'Ace', suit: 'Spades')
```

instead of 

```
card = Card.new('Ace', 'Spades')
```

The former is more code, but it's a little more expressive.  It would open up opportunities like this:

```
card = Card.new(joker: 'Red')
# instead of card = Card.new(nil, nil, 'Red')
```

I think Jokers could be considered as a rank, but... they don't really fall in the rank/suit concept, so why force it?  Internally the joker could be handled as a rank, or... done separately, but the point is:  We get around this concept flip, and we get more expressive code.  But at the cost of breaking changes.  But hey, the library's new, right? :smile:

File: lib/rubycards/deck.rb [hunk 1]
```diff
@@ -17,13 +17,19 @@ class Deck
...
     # @return [Deck] A standard deck of cards
-    def initialize
+    def initialize(options = {})
       @cards = []
 
       RANKS.product(SUITS).each do |rank, suit|
         @cards << Card.new(rank, suit)
       end
+
+      if options.any?
```
Comment (joeletizia) [inline lib/rubycards/deck.rb]: Maybe we should change this to 

``` ruby
if options[:joker]
```

**Comment by joeletizia** [comment]
Also, for the constructors, we could do

``` ruby
Card.new(nil, nil, joker: true)
Deck.new(joker: true)
```

File: lib/rubycards/deck.rb (hunk 1 above)
Comment (darrencauthon) [inline lib/rubycards/deck.rb]: Perhaps, yes... I didn't get that far.  I was in full TDD mode, and I just haven't gotten to the point where I needed to write a test that made the blind `options.any?`

Maybe I could post this branch to the main project, and then anybody could work on it?

**Comment by darrencauthon** [comment]
@joeletizia Hmm, I agree, that's a good call.  I like `jokers` better than `include_jokers` 

I only know games with two jokers, but after some wikipedia research I saw that there are games that have one joker.  So perhaps we could also allow `Deck.new(joker: 1)` for one joker, `Deck.new(joker: 2)` for two, and `Deck.new(joker: true)` for two jokers (which I believe is the more common use).

**Comment by jdan** [comment]
Let's roll with this:

``` ruby
Deck.new(jokers: true)    # two jokers
Deck.new(jokers: 1)

Card.new('Ace', 'Spades')
Card.new('Joker')
Card.new('Joker', 'Red')
```

Seem plausible?

**Comment by joeletizia** [comment]
There's games that use one joker?

**Comment by darrencauthon** [comment]
@joeletizia Yeah, there are games that only use one joker.

I did some initial research before I even started this, and I was surprised how whishy-washy the state of the joker is.  The regular card deck is very consistent, but the jokers vary wildly.... even with what the jokers are called.

I think I saw that, overall, a Red/Black seemed consistent, with one serving as a "high" and one serving as a "low."  _shrug_ It can be hard. :smile:

File: lib/rubycards/deck.rb (hunk 1 above)
Comment (darrencauthon) [inline lib/rubycards/deck.rb]: @joeletizia I just wanted to point out one thing, here:

https://github.com/darrencauthon/rubycards/commit/e6c424f5b59b87129220c5b00a0396a6a92d0e93

When I merged in the changes from master, tests failed... which forced me to update the code like you wanted.

I know, we can eye-ball these sorts of things and say, "Oh, I'm passing in this, and I'm checking against this... but I **love** having a test force to to explicitly state that. :smile:

**Comment by darrencauthon** [comment]
I've been really busy, and just haven't gotten back to this.

I've pushed this branch into the main project, here:  https://github.com/jdan/rubycards/tree/jokers

In case anybody wants to finish this without me.  I'll get to it _eventually_....

**Comment by darrencauthon** [comment]
I'm not sure why this suddenly appeared in PR list... today, in 2023!  

I enjoyed working on this project. 😄

**Pull request closed without merging**